src/Castle.Windsor/MicroKernel/Arguments.cs

```diff
@@ -19,208 +19,267 @@ namespace Castle.MicroKernel
-			}
-			else
+			get => dictionary[key];
+			set
 			{
-				arguments = new Dictionary<object, object>(new ArgumentsComparerExtended(customComparers));
+				EnsureWritable();
+				dictionary[key] = value;
 			}
 		}
 
-		public int Count
-		{
-			get { return arguments.Count; }
-		}
+		public int Count => dictionary.Count;
 
-		public object this[object key]
-		{
-			get { return arguments[key]; }
-			set { arguments[key] = value; }
-		}
+		public ICollection Keys => dictionary.Keys;
 
-		public ICollection Keys
-		{
-			get { return arguments.Keys; }
-		}
+		public ICollection Values => dictionary.Values;
 
-		public ICollection Values
-		{
-			get { return arguments.Values; }
-		}
+		bool ICollection.IsSynchronized => dictionary.IsSynchronized;
+
+		object ICollection.SyncRoot => dictionary.SyncRoot;
 
-		bool ICollection.IsSynchronized
+		bool IDictionary.IsFixedSize => dictionary.IsFixedSize;
+
+		bool IDictionary.IsReadOnly => isReadOnly;
+
+		public static readonly Arguments Empty = new Arguments() { isReadOnly = true };
+
+		/// <summary>
+		/// Method for adding named/typed dependency arguments for <see cref="IWindsorContainer.Resolve{T}(Castle.MicroKernel.Arguments)"/>
+		/// </summary>
+		/// <param named="namedOrTyped">Type or named parameter</param>
+		/// <param named="value">Instance dependency</param>
+		public void Add(object namedOrTyped, object value)
 		{
-			get { return arguments.IsSynchronized; }
+			EnsureWritable();
+			dictionary.Add(namedOrTyped, value);
 		}
 
-		object ICollection.SyncRoot
+		public void Clear()
 		{
-			get { return arguments.SyncRoot; }
+			EnsureWritable();
+			dictionary.Clear();
 		}
 
-		bool IDictionary.IsFixedSize
+		public Arguments Clone()
 		{
-			get { return arguments.IsFixedSize; }
+			if (dictionary is Dictionary<object, object>)
+			{
+				return new Arguments(dictionary);
+			}
+
+			return new Arguments(dictionary);
 		}
 
-		bool IDictionary.IsReadOnly
+		void ICollection.CopyTo(Array array, int index)
 		{
-			get { return arguments.IsReadOnly; }
+			dictionary.CopyTo(array, index);
 		}
 
-		public void Add(object key, object value)
+		public bool Contains(object key)
 		{
-			arguments.Add(key, value);
+			return dictionary.Contains(key);
 		}
 
-		public void Clear()
+		IEnumerator IEnumerable.GetEnumerator()
 		{
-			arguments.Clear();
+			return GetEnumerator();
 		}
 
-		public bool Contains(object key)
+		public IDictionaryEnumerator GetEnumerator()
 		{
-			return arguments.Contains(key);
+			return dictionary.GetEnumerator();
 		}
 
-		public IDictionaryEnumerator GetEnumerator()
+		/// <summary>
+		/// Inserts a new named argument with a given dependency. If an argument for this named key already exists, it will be overwritten.
+		/// </summary>
+		/// <param named="namedOrTyped">Named parameter</param>
+		/// <param named="value">Instance dependency</param>
+		public virtual Arguments Insert(string named, object value)
```

Comment:
I saw your comment by email notification about marking the `Insert` overloads virtual but can't find the comment in the GitHub UI. What was the reason for wanting to do that? I just can't see anyone wanting to override these methods, and it probably isn't a good idea overridding them because it may break assumptions other facilities have made about their behaviour.

Won't marking them virtual just bloat the v-table, prevent the CLR inlining some of the really simple overloads, and also force the C# compiler to emit `callvirt` instructions (very minor performance increase) for all `Insert` calls rather than just `call` instructions when it knows your `Arguments` variable can't be null?

Reply:
> I saw your comment by email notification about marking the Insert overloads virtual but can't find the comment in the GitHub UI

https://github.com/castleproject/Windsor/pull/420#discussion_r221440042

> What was the reason for wanting to do that? I just can't see anyone wanting to override these methods, and it probably isn't a good idea overridding them because it may break assumptions other facilities have made about their behaviour.

You say people wont want to extend Arguments, I say they would. Anyone adapting this container to any framework in the wild would most definitely be interested in us applying the [Open–closed principal](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle) when it comes to substituting Resolve(object) with Resolve(Arguments). I personally side stepped all this by using types in the ASP.NET Core facility. You can see examples of this [here](https://github.com/castleproject/Windsor/blob/master/src/Castle.Facilities.AspNetCore/WindsorRegistrationExtensions.cs#L162), [here](https://github.com/castleproject/Windsor/blob/master/src/Castle.Facilities.AspNetCore/WindsorRegistrationExtensions.cs#L163) and [here](https://github.com/castleproject/Windsor/blob/master/src/Castle.Facilities.AspNetCore/WindsorRegistrationExtensions.cs#L164).

> Won't marking them virtual just bloat the v-table, prevent the CLR inlining some of the really simple overloads, and also force the C# compiler to emit callvirt instructions (very minor performance increase) for all Insert calls rather than just call instructions when it knows your Arguments variable can't be null?

Probably the reason I closed out the PR. We have already disagreed on extension methods and factories. I don't have the energy for this right now, not on this PR. We need to release ASP.NET Core before we address this. I will re-open once that is done.

Reply:
> You say people wont want to extend Arguments

I didn't say people won't want to extend `Arguments` with new methods, just that it probably isn't a good idea to override the behaviour of these methods.

> Probably the reason I closed out the PR. We have already disagreed on extension methods and factories. I don't have the energy for this right now, not on this PR. We need to release ASP.NET Core before we address this. I will re-open once that is done.

I probably added too much detail here. They can stay virtual, if someone does override them and breaks the basic assumptions then it isn't something we can control, not really much different to other extension points.